### DISCUSSION TITLE
Payload CLI

### DISCUSSION THREAD

**Discussion by paulpopus** [payloadcms/payload#2505] [Feature Requests & Ideas]
I think it would be very useful to start building a CLI for interacting with Payload to make configuration and setup of various features a lot smoother. It also opens the gate for future functions or server-side automation without using an API endpoint.

# Possible features

## Global
- Install payload from templates

## Project level
- **Install plugins** - somewhat related to this issue https://github.com/payloadcms/payload/discussions/2504
- Generate code for new collections/globals/fields
- Generate endpoints for REST/GraphQL API
- Generate a one time login link (drupal does this and it's great for dev exp on the server)
- Run registered commands by the developer for other types of automation

# Proposed plan

We would start with a baseline of features, most useful from a developer's perspective such as code gen. Generally I'm looking for other people's thoughts on this so we can actually flesh out the idea a bit more.

**Comment by TimHal** [comment]
Love the proposal and was thinking of the same. I would even go so far to propose a runtime-cli. One major benefit would be that you can hide the admin dashboard (even get rid of react?) and still be able to ssh into your server and quickly edit a document.

**Comment by sriechersrc** [comment]
It would be really nice to have some configuration options to choose from, kind of like [T3 App](https://create.t3.gg/) does it. The templates at the moment are a little bit too specialized on a particular use case. I would love to have the option to generate a website template but with payments and configured for deployment on vercel for example. Right now I need to rip out all the parts I don't need manually

**Comment by lstellway** [comment]
What brought me here is starting to play with [Pi](https://pi.dev):

> _[It intentionally does not include built-in MCP, sub-agents, permission popups, plan mode, to-dos, or background bash [...]](https://pi.dev/docs/latest/usage#design-principles)_

Definitely a bit disorienting after spending the past few months getting accustomed with Claude / MCP. 

But the move is turning out to be pretty simple for a couple tools I'm coming across:

- [herdrdev/herdr](https://github.com/herdrdev/herdr) ships with its own [skills](https://github.com/herdrdev/herdr/tree/master/skills)
- [nesszer/linear-cli](https://github.com/nesszer/linear-cli) ships with its own [skills](https://github.com/nesszer/linear-cli/tree/master/skills)

You just "teach it" to use the CLI tools you already have _(give it context via skills)_.

We already have the [Payload MCP](https://payloadcms.com/docs/plugins/mcp) plugin.
I would LOVE to be able to interact with my Payload instance via CLI. 

Would make sense to nail down a scope.. 
I'd personally be happy just being able to manage _(CRUD operations)_ my collections.

**Comment by lstellway** [comment]
Seems like this is a feature slotted in for v4!

https://github.com/payloadcms/payload/blob/e0be88d56010fb6ea1e12ffa2c609df63499eaa1/docs/configuration/cli.mdx
